parse.ml

```diff
@@ -9,9 +9,44 @@ let split_annotations program =
   let instructions = Array.map snd program in
   (instructions, annotations)
 
+exception Parse_error of string option * Lexing.position * Lexing.position
+
+let parse lexbuf =
+  (* see the Menhir manual for the description of
+     error messages support *)
+  let open MenhirLib.General in
+  let module Interp = Parser.MenhirInterpreter in
+  let input = Interp.lexer_lexbuf_to_supplier Lexer.token lexbuf in
+  let success prog = split_annotations prog in
+  let failure error_state =
+    let env = match error_state with
+      | Interp.HandlingError env -> env
+      | _ -> assert false in
+    match Interp.stack env with
+    | lazy Nil -> assert false
+    | lazy (Cons (Interp.Element (state, _, start_pos, end_pos), _)) ->
+      let message =
+        try Some (Parser_messages.message (Interp.number state)) with
+        | Not_found -> None in
+      raise (Parse_error (message, start_pos, end_pos))
+  in
+  Interp.loop_handle success failure input
+    (Parser.Incremental.program lexbuf.Lexing.lex_curr_p)
+
 let parse_string str =
   let lexbuf = Lexing.from_string str in
-  split_annotations (Parser.program Lexer.token lexbuf)
+  parse lexbuf
```

Comment:
I know it's actually my leftovers -- but if you don't mind it would be nice to somehow have better errors for parse_string too